server/jetstream_cluster_test.go

```diff
@@ -5385,27 +5366,58 @@ func TestJetStreamClusterMirrorAndSourcesClusterRestart(t *testing.T) {
 		}
 	}
 
-	checkSync := func() {
+	checkSync := func(msgsTest, msgsM uint64) {
 		t.Helper()
 		checkFor(t, 10*time.Second, 500*time.Millisecond, func() error {
-			tsi, err := js.StreamInfo("TEST")
-			if err != nil {
-				t.Fatalf("Unexpected error: %v", err)
-			}
-			msi, err := js.StreamInfo("M")
-			if err != nil {
-				t.Fatalf("Unexpected error: %v", err)
-			}
-			if tsi.State.Msgs != msi.State.Msgs {
-				return fmt.Errorf("Total messages not the same: TEST %d vs M %d", tsi.State.Msgs, msi.State.Msgs)
+			if tsi, err := js.StreamInfo("TEST"); err != nil {
+				return err
+			} else if msi, err := js.StreamInfo("M"); err != nil {
+				return err
+			} else if tsi.State.Msgs != msgsTest {
+				return fmt.Errorf("received %d msgs from TEST, expected %d", tsi.State.Msgs, msgsTest)
+			} else if msi.State.Msgs != msgsM {
+				return fmt.Errorf("received %d msgs from TEST, expected %d", msi.State.Msgs, msgsM)
 			}
 			return nil
 		})
 	}
 
+	// Origin
+	_, err := js.AddStream(&nats.StreamConfig{
+		Name:     "TEST",
+		Subjects: []string{"foo"},
+		Replicas: 2,
+	})
+	require_NoError(t, err)
+
+	_, err = js.AddStream(&nats.StreamConfig{
+		Name:     "M",
+		Sources:  []*nats.StreamSource{{Name: "TEST", FilterSubject: "foo"}},
+		Replicas: 2,
+	})
+
+	require_NoError(t, err)
+
 	// Send 100 msgs.
 	sendBatch("foo", 100)
-	checkSync()
+	checkSync(100, 100)
+
+	// update makes source invalid
+	_, err = js.UpdateStream(&nats.StreamConfig{
+		Name:     "TEST",
+		Subjects: []string{"bar"},
+		Replicas: 2,
+	})
+	require_NoError(t, err)
+
+	// TODO check for downstream error propagation
+
+	_, err = js.Publish("foo", nil)
+	require_Error(t, err)
+
+	sendBatch("bar", 100)
+	// The source stream remains at 100 msgs as it still uses foo as it's filter
+	checkSync(200, 100)
```

Comment:
after this line: call nc.Close() here otherwise the original connection would try to reconnect, which can delay things or affect other tests, etc...

Reply:
done